# Advanced MLP - 2
- Putting it altogether

In [11]:
import matplotlib.pyplot as plt
import numpy as np

from sklearn.ensemble import VotingClassifier
from sklearn.metrics import accuracy_score
from sklearn.model_selection import train_test_split
from tensorflow.keras.wrappers.scikit_learn import KerasClassifier

from tensorflow.keras.models import Sequential
from tensorflow.keras.utils import to_categorical
from tensorflow.keras.layers import Activation, Dense, BatchNormalization, Dropout
from tensorflow.keras import optimizers

from tensorflow.keras.datasets import mnist

## Load Dataset

In [2]:
(X_train, y_train), (X_test, y_test) = mnist.load_data()

In [3]:
# reshaping X data: (n, 28, 28) => (n, 784)
X_train = X_train.reshape((X_train.shape[0], X_train.shape[1] * X_train.shape[2]))
X_test = X_test.reshape((X_test.shape[0], X_test.shape[1] * X_test.shape[2]))

In [4]:
# We use all training data and validate on all test data
print(X_train.shape)
print(X_test.shape)
print(y_train.shape)
print(y_test.shape)

(60000, 784)
(10000, 784)
(60000,)
(10000,)


## Training & Validating Model
- Measures to improve training is applied simultaneously
    - More training set
    - Weight Initialization scheme
    - Nonlinearity (Activation function)
    - Optimizers: adaptvie
    - Batch Normalization
    - Dropout (Regularization)
    - Model Ensemble

In [5]:
def mlp_model():
    model = Sequential()
    
    model.add(Dense(50, input_shape = (784, ), kernel_initializer='he_normal'))
    model.add(BatchNormalization())
    model.add(Activation('relu'))
    model.add(Dropout(0.2))
    model.add(Dense(50, kernel_initializer='he_normal'))
    model.add(BatchNormalization())
    model.add(Activation('relu'))    
    model.add(Dropout(0.2))
    model.add(Dense(50, kernel_initializer='he_normal'))
    model.add(BatchNormalization())
    model.add(Activation('relu'))
    model.add(Dropout(0.2))
    model.add(Dense(50, kernel_initializer='he_normal'))
    model.add(BatchNormalization())
    model.add(Activation('relu'))
    model.add(Dropout(0.2))
    model.add(Dense(10, kernel_initializer='he_normal'))
    model.add(Activation('softmax'))
    
    adam = optimizers.Adam(lr = 0.001)
    model.compile(optimizer = adam, loss = 'categorical_crossentropy', metrics = ['acc'])
    
    return model

In [6]:
# create 5 models to ensemble
model1 = KerasClassifier(build_fn = mlp_model, epochs = 100)
model2 = KerasClassifier(build_fn = mlp_model, epochs = 100)
model3 = KerasClassifier(build_fn = mlp_model, epochs = 100)
model4 = KerasClassifier(build_fn = mlp_model, epochs = 100)
model5 = KerasClassifier(build_fn = mlp_model, epochs = 100)

In [7]:
ensemble_clf = VotingClassifier(estimators = [('model1', model1), ('model2', model2), ('model3', model3), ('model4', model4), ('model5', model5)], voting = 'soft')

In [8]:
ensemble_clf.fit(X_train, y_train)

Train on 60000 samples
Epoch 1/100
60000/60000 [==============================] - 6s 95us/sample - loss: 0.7740 - acc: 0.7621
Epoch 2/100
60000/60000 [==============================] - 5s 80us/sample - loss: 0.4339 - acc: 0.8753
Epoch 3/100
60000/60000 [==============================] - 5s 82us/sample - loss: 0.3631 - acc: 0.8970
Epoch 4/100
60000/60000 [==============================] - 5s 84us/sample - loss: 0.3250 - acc: 0.9071
Epoch 5/100
60000/60000 [==============================] - 4s 73us/sample - loss: 0.3032 - acc: 0.9131
Epoch 6/100
60000/60000 [==============================] - 4s 73us/sample - loss: 0.2816 - acc: 0.9197
Epoch 7/100
60000/60000 [==============================] - 5s 77us/sample - loss: 0.2686 - acc: 0.9227
Epoch 8/100
60000/60000 [==============================] - 5s 76us/sample - loss: 0.2571 - acc: 0.9267
Epoch 9/100
60000/60000 [==============================] - 5s 77us/sample - loss: 0.2465 - acc: 0.9302
Epoch 10/100
60000/60000 [========================

60000/60000 [==============================] - 4s 72us/sample - loss: 0.1500 - acc: 0.9558
Epoch 80/100
60000/60000 [==============================] - 4s 73us/sample - loss: 0.1503 - acc: 0.9564
Epoch 81/100
60000/60000 [==============================] - 4s 72us/sample - loss: 0.1498 - acc: 0.9568
Epoch 82/100
60000/60000 [==============================] - 4s 73us/sample - loss: 0.1480 - acc: 0.9569
Epoch 83/100
60000/60000 [==============================] - 4s 73us/sample - loss: 0.1505 - acc: 0.9564
Epoch 84/100
60000/60000 [==============================] - 4s 72us/sample - loss: 0.1503 - acc: 0.9552
Epoch 85/100
60000/60000 [==============================] - 4s 73us/sample - loss: 0.1523 - acc: 0.9560
Epoch 86/100
60000/60000 [==============================] - 4s 72us/sample - loss: 0.1488 - acc: 0.9557
Epoch 87/100
60000/60000 [==============================] - 4s 72us/sample - loss: 0.1457 - acc: 0.9574
Epoch 88/100
60000/60000 [==============================] - 4s 72us/sample - 

60000/60000 [==============================] - 4s 74us/sample - loss: 0.1618 - acc: 0.9533
Epoch 58/100
60000/60000 [==============================] - 5s 75us/sample - loss: 0.1563 - acc: 0.9543
Epoch 59/100
60000/60000 [==============================] - 4s 74us/sample - loss: 0.1578 - acc: 0.9537
Epoch 60/100
60000/60000 [==============================] - 5s 77us/sample - loss: 0.1612 - acc: 0.9532
Epoch 61/100
60000/60000 [==============================] - 4s 74us/sample - loss: 0.1550 - acc: 0.9545
Epoch 62/100
60000/60000 [==============================] - 5s 75us/sample - loss: 0.1587 - acc: 0.9525
Epoch 63/100
60000/60000 [==============================] - 4s 74us/sample - loss: 0.1576 - acc: 0.9539
Epoch 64/100
60000/60000 [==============================] - 4s 74us/sample - loss: 0.1567 - acc: 0.9536
Epoch 65/100
60000/60000 [==============================] - 4s 72us/sample - loss: 0.1549 - acc: 0.9548
Epoch 66/100
60000/60000 [==============================] - 4s 70us/sample - 

60000/60000 [==============================] - 4s 68us/sample - loss: 0.1770 - acc: 0.9498
Epoch 36/100
60000/60000 [==============================] - 4s 68us/sample - loss: 0.1762 - acc: 0.9487
Epoch 37/100
60000/60000 [==============================] - 4s 68us/sample - loss: 0.1717 - acc: 0.9509
Epoch 38/100
60000/60000 [==============================] - 4s 68us/sample - loss: 0.1727 - acc: 0.9502
Epoch 39/100
60000/60000 [==============================] - 4s 68us/sample - loss: 0.1684 - acc: 0.9512
Epoch 40/100
60000/60000 [==============================] - 4s 68us/sample - loss: 0.1682 - acc: 0.9512
Epoch 41/100
60000/60000 [==============================] - 4s 68us/sample - loss: 0.1671 - acc: 0.9510
Epoch 42/100
60000/60000 [==============================] - 4s 68us/sample - loss: 0.1653 - acc: 0.9518
Epoch 43/100
60000/60000 [==============================] - 7s 109us/sample - loss: 0.1692 - acc: 0.9507
Epoch 44/100
60000/60000 [==============================] - 5s 78us/sample -

60000/60000 [==============================] - 4s 73us/sample - loss: 0.2215 - acc: 0.9362
Epoch 14/100
60000/60000 [==============================] - 4s 73us/sample - loss: 0.2202 - acc: 0.9373
Epoch 15/100
60000/60000 [==============================] - 4s 73us/sample - loss: 0.2199 - acc: 0.9367
Epoch 16/100
60000/60000 [==============================] - 4s 73us/sample - loss: 0.2117 - acc: 0.9383
Epoch 17/100
60000/60000 [==============================] - 4s 73us/sample - loss: 0.2041 - acc: 0.9420
Epoch 18/100
60000/60000 [==============================] - 4s 72us/sample - loss: 0.2041 - acc: 0.9406
Epoch 19/100
60000/60000 [==============================] - 4s 72us/sample - loss: 0.2029 - acc: 0.9412
Epoch 20/100
60000/60000 [==============================] - 4s 73us/sample - loss: 0.1968 - acc: 0.9432
Epoch 21/100
60000/60000 [==============================] - 4s 72us/sample - loss: 0.1977 - acc: 0.9427
Epoch 22/100
60000/60000 [==============================] - 5s 75us/sample - 

60000/60000 [==============================] - 4s 67us/sample - loss: 0.1446 - acc: 0.9585
Epoch 92/100
60000/60000 [==============================] - 4s 68us/sample - loss: 0.1426 - acc: 0.9585
Epoch 93/100
60000/60000 [==============================] - 4s 68us/sample - loss: 0.1420 - acc: 0.9581
Epoch 94/100
60000/60000 [==============================] - 4s 68us/sample - loss: 0.1462 - acc: 0.9566
Epoch 95/100
60000/60000 [==============================] - 4s 68us/sample - loss: 0.1436 - acc: 0.9572
Epoch 96/100
60000/60000 [==============================] - 4s 68us/sample - loss: 0.1446 - acc: 0.9577
Epoch 97/100
60000/60000 [==============================] - 4s 68us/sample - loss: 0.1425 - acc: 0.9586
Epoch 98/100
60000/60000 [==============================] - 4s 69us/sample - loss: 0.1442 - acc: 0.9575
Epoch 99/100
60000/60000 [==============================] - 4s 68us/sample - loss: 0.1448 - acc: 0.9574
Epoch 100/100
60000/60000 [==============================] - 4s 67us/sample -

60000/60000 [==============================] - 4s 68us/sample - loss: 0.1516 - acc: 0.9561
Epoch 70/100
60000/60000 [==============================] - 4s 68us/sample - loss: 0.1562 - acc: 0.9537
Epoch 71/100
60000/60000 [==============================] - 4s 68us/sample - loss: 0.1525 - acc: 0.9562
Epoch 72/100
60000/60000 [==============================] - 4s 68us/sample - loss: 0.1506 - acc: 0.9561
Epoch 73/100
60000/60000 [==============================] - 4s 68us/sample - loss: 0.1524 - acc: 0.9553
Epoch 74/100
60000/60000 [==============================] - 4s 69us/sample - loss: 0.1515 - acc: 0.9558
Epoch 75/100
60000/60000 [==============================] - 4s 69us/sample - loss: 0.1514 - acc: 0.9556
Epoch 76/100
60000/60000 [==============================] - 4s 68us/sample - loss: 0.1530 - acc: 0.9560
Epoch 77/100
60000/60000 [==============================] - 4s 68us/sample - loss: 0.1511 - acc: 0.9566
Epoch 78/100
60000/60000 [==============================] - 4s 68us/sample - 

VotingClassifier(estimators=[('model1', <tensorflow.python.keras.wrappers.scikit_learn.KerasClassifier object at 0x141bcb240>), ('model2', <tensorflow.python.keras.wrappers.scikit_learn.KerasClassifier object at 0x141bcb278>), ('model3', <tensorflow.python.keras.wrappers.scikit_learn.KerasClassifier object at 0x141b..., ('model5', <tensorflow.python.keras.wrappers.scikit_learn.KerasClassifier object at 0x141bcb358>)],
         flatten_transform=None, n_jobs=None, voting='soft', weights=None)

In [9]:
y_pred = ensemble_clf.predict(X_test)

In [10]:
print('Acc: ', accuracy_score(y_pred, y_test))

Acc:  0.9794


Final model produces over **98% of accuracy**, which is rather improved from before